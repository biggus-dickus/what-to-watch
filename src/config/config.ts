// How much time to wait in ms before playing a film preview
export const HOVER_DELAY = 1000;

// How many films to render in "More like this" section (film page)
export const MORE_LIKE_THIS_LIMIT = 4;

// How many films to display at once on mount
export const FILMS_LIMIT = 20;

// How many genres to display at once
export const GENRES_LIMIT = 9 + 1; // 9 + "All genres"

// Navigation tabs on film page
export const NAV_OVERVIEW_ID = `overview`;
export const NAV_DETAILS_ID = `details`;
export const NAV_REVIEWS_ID = `reviews`;
export const filmTabs = [
  {
    id: NAV_OVERVIEW_ID,
    text: `Overview`
  },
  {
    id: NAV_DETAILS_ID,
    text: `Details`
  },
  {
    id: NAV_REVIEWS_ID,
    text: `Reviews`
  }
];

// Film review constraints
export const REVIEW_CHARS_MIN = 50;
export const REVIEW_CHARS_MAX = 400;
