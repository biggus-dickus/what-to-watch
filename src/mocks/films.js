export const films = [
  {
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    picUrl: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Bohemian Rhapsody`,
    picUrl: `img/bohemian-rhapsody.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Macbeth`,
    picUrl: `img/macbeth.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Aviator`,
    picUrl: `img/aviator.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `We need to talk about Kevin`,
    picUrl: `img/we-need-to-talk-about-kevin.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `What We Do in the Shadows`,
    picUrl: `img/what-we-do-in-the-shadows.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Revenant`,
    picUrl: `img/revenant.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Johny English`,
    picUrl: `img/johnny-english.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Shutter Island`,
    picUrl: `img/shutter-island.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Pulp Fiction`,
    picUrl: `img/pulp-fiction.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `No Country for Old Men`,
    picUrl: `img/no-country-for-old-men.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Snatch`,
    picUrl: `img/snatch.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Moonrise Kingdom`,
    picUrl: `img/moonrise-kingdom.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Seven Years in Tibet`,
    picUrl: `img/seven-years-in-tibet.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Midnight Special`,
    picUrl: `img/midnight-special.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `War of the Worlds`,
    picUrl: `img/war-of-the-worlds.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Dardjeeling Limited`,
    picUrl: `img/dardjeeling-limited.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Orlando`,
    picUrl: `img/orlando.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
  {
    name: `Mindhunter`,
    picUrl: `img/mindhunter.jpg`,
    pageUrl: `https://thepiratebay.org`
  },
];
