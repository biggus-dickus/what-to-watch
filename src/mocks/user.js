const mockUser = {
  id: 1,
  email: `Oliver.conner@gmail.com`,
  name: `Oliver.conner`,
  avatar_url: `img/1.png` // eslint-disable-line
};

export default mockUser;
