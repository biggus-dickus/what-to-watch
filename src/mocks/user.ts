import {Location, User} from '../types'; // eslint-disable-line

export const mockLocation: Location = {
  hash: `#p1t00h`,
  key: `0mfg666`,
  pathname: `/`,
  search: ``
};

const mockUser: User = {
  avatar_url: '/wtw/static/avatar/10.jpg', // eslint-disable-line
  id: 777,
  email: `test@test.com`,
  name: `Nastasha Romanenko`,
};

export default mockUser;
