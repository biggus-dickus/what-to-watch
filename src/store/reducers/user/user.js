import {ActionType} from '../../action-types';

const initialState = {
  isAuthorizationRequired: false,
  userData: null
};

export const userReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case ActionType.AUTH_REQUIRED:
      return {
        ...state,
        isAuthorizationRequired: action.payload
      };

    case ActionType.GET_USER_DATA:
      return {
        ...state,
        userData: action.payload
      };

    default: return state;
  }
};
