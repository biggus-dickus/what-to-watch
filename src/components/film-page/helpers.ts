import {Film} from '../../types'; // eslint-disable-line
import {MORE_LIKE_THIS_LIMIT} from '../../config/config';


export const getFilmById = (films: Film[], id: number): Film | undefined => films.find((film) => film.id === id);

export const getMoreLikeThis = (films: Film[], genre: string, notId: number): Film[] | [] => {
  const result = films.filter((film) => film.genre === genre && film.id !== notId);

  return result.slice(0, MORE_LIKE_THIS_LIMIT);
};
