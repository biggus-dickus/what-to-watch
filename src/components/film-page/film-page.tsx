import * as React from 'react';
import {connect} from 'react-redux';

import {Film, Location, User} from '../../types'; // eslint-disable-line
import {NAV_OVERVIEW_ID, NAV_DETAILS_ID, NAV_REVIEWS_ID, filmTabs} from '../../config/config';

import {getError, getReviews} from '../../store/reducers/data/selectors';
import {getFilmById, getMoreLikeThis} from './helpers';
import {Operation} from '../../store/operations';

import FilmButtons from './film-buttons/film-buttons';
import FilmDetails from './film-details/film-details';
import FilmsList from '../films-list/films-list';
import FilmNav from './film-nav/film-nav';
import FilmOverview from './film-overview/film-overview';
import FilmReviews from './film-reviews/film-reviews';
import FilmTitle from './film-title/film-title';
import Footer from '../partials/footer/footer';
import Logo from '../partials/logo/logo';
import NoMatch from '../no-match/no-match';
import UserBlock from '../partials/user-block/user-block';

interface Props {
  availableMovies: Film[],
  computedMatch: any,
  error?: string,
  location: Location,
  onAddToWatchList: (id: number) => Promise<any>,
  onRemoveFromWatchList: (id: number) => Promise<any>,
  reviews?: any[],
  userData?: User,
  loadReviews: (filmId: number) => void
}

const getValidTabId = (hashVal:string):string => {
  if (hashVal && [NAV_OVERVIEW_ID, NAV_DETAILS_ID, NAV_REVIEWS_ID].includes(hashVal)) {
    return hashVal;
  }

  return NAV_OVERVIEW_ID;
};


export const FilmPage = (props: Props): React.ReactElement => {
  const {
    availableMovies,
    computedMatch,
    location,
    onAddToWatchList,
    onRemoveFromWatchList,
    userData,
    loadReviews
  } = props;

  const {hash, pathname} = location;

  const [activeTabId, setActiveTabId] = React.useState(getValidTabId(hash.split(`#`)[1]));

  const [didMount, setDidMount] = React.useState(false);
  React.useEffect(() => setDidMount(true), []);

  // Fetch reviews on load and subsequent location changes
  React.useEffect(() => {
    (async function () {
      await loadReviews(+computedMatch.params.id);
    })();
  }, [pathname]);

  // Revert active tab to default one, but only on subsequent changes to location
  React.useEffect(() => {
    if (didMount) {
      setActiveTabId(NAV_OVERVIEW_ID);
    }
  }, [pathname]);

  const handleTabChange = (newTabId: string): void => {
    if (activeTabId !== newTabId) {
      setActiveTabId(newTabId);
    }
  };

  const film = getFilmById(availableMovies, +computedMatch.params.id);

  if (film) {
    const similarFilms = getMoreLikeThis(availableMovies, film.genre, film.id);

    return (
      <>
        <section
          className="movie-card movie-card--full"
          style={{background: film.bgColor}}>
          <div className="movie-card__hero">
            <div className="movie-card__bg">
              <img src={film.bgImage} alt={film.name} />
            </div>

            <header className="page-header movie-card__head">
              <Logo pathname={location.pathname} />

              <UserBlock location={location} user={userData} />
            </header>

            <div className="movie-card__wrap">
              <div className="movie-card__desc">
                <FilmTitle
                  name={film.name}
                  genre={film.genre}
                  released={film.released}
                  ish1 />

                <FilmButtons
                  filmId={film.id}
                  isAdded={film.isFavourite}
                  {...{onAddToWatchList, onRemoveFromWatchList, location}}
                />
              </div>
            </div>
          </div>

          <div className="movie-card__wrap movie-card__translate-top">
            <div className="movie-card__info">
              <div className="movie-card__poster movie-card__poster--big">
                <img src={film.posterImage} alt={`${film.name} poster`} />
              </div>

              <div className="movie-card__desc">
                <FilmNav
                  tabs={filmTabs}
                  activeTabId={activeTabId}
                  onTabClick={handleTabChange} />

                {activeTabId === NAV_OVERVIEW_ID && (
                  <FilmOverview
                    description={film.description}
                    director={film.director}
                    rating={film.rating}
                    scoresCount={film.scoresCount}
                    starring={film.starring} />
                )}

                {activeTabId === NAV_DETAILS_ID && (
                  <FilmDetails
                    director={film.director}
                    genre={film.genre}
                    released={film.released}
                    runTime={film.runTime}
                    starring={film.starring} />
                )}

                {activeTabId === NAV_REVIEWS_ID && (
                  <FilmReviews
                    error={props.error}
                    filmId={film.id}
                    reviews={props.reviews} />
                )}
              </div>
            </div>
          </div>
        </section>

        <div className="page-content">
          {similarFilms.length ? (
            <section className="catalog catalog--like-this">
              <h2 className="catalog__title">More like this</h2>
              <FilmsList films={similarFilms} />
            </section>
          ) : null}

          <Footer pathname={location.pathname} />
        </div>
      </>
    );
  }

  return <NoMatch />;
};

const mapStateToProps = (state) => ({
  error: getError(state),
  reviews: getReviews(state)
});

const mapDispatchToProps = (dispatch) => ({
  loadReviews: (id) => dispatch(Operation.fetchReviews(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(FilmPage);
