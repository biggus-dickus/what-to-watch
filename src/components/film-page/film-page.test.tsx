import * as React from 'react';
import {StaticRouter as Router} from 'react-router';

import * as renderer from 'react-test-renderer';
import * as Adapter from 'enzyme-adapter-react-16';
import {configure, shallow} from 'enzyme';

import {mockFilms} from '../../mocks/films';
import {mockLocation} from '../../mocks/user';
import {NAV_DETAILS_ID} from '../../config/config';

import {getFilmById, getMoreLikeThis} from './helpers';
import {FilmPage} from './film-page';

configure({adapter: new Adapter()});


const props = {
  availableMovies: mockFilms,
  computedMatch: {params: {id: 2}},
  location: {...mockLocation, pathname: `/film/2`},
  loadReviews: jest.fn()
};

const film = getFilmById(mockFilms, props.computedMatch.params.id);
const similarFilms = getMoreLikeThis(mockFilms, film.genre, film.id);


describe(`FilmPage test suite`, () => {
  it(`<FilmPage /> should render correctly`, () => {
    const tree = renderer.create(<Router><FilmPage {...props} /></Router>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should render a 404 page if the passed id is not within the range of available film IDs`, () => {
    const newProps = {...props, computedMatch: {params: {id: 666}}};

    const wrapper = shallow(<FilmPage {...newProps} />);
    expect(wrapper.find(`NoMatch`)).toHaveLength(1);
  });

  it(`should render "More like this" (films of the same genre) correctly`, () => {
    const wrapper = shallow(<FilmPage {...props} />);
    const filmsList = wrapper.find(`FilmsList`);

    expect(filmsList.prop(`films`)).toHaveLength(similarFilms.length);
  });

  it(`should NOT contain the currently displayed film in the list of similar films`, () => {
    const res = similarFilms.find((it) => it.id === film.id);
    expect(res).toBeUndefined();
  });

  it(`should NOT render "More like this" section if there are no more films of selected genre`, () => {
    const newProps = {
      ...props,
      computedMatch: {params: {id: 1}},
      location: {...props.location, pathname: `/film/1`}
    };

    const wrapper = shallow(<FilmPage {...newProps} />);
    const filmsList = wrapper.find(`FilmsList`);

    expect(filmsList).toHaveLength(0);
  });

  it(`should NOT render more than 4 films in "More like this section"`, () => {
    const newProps = {
      ...props,
      computedMatch: {params: {id: 3}}, // Drama
      location: {...props.location, pathname: `/film/3`}
    };

    const wrapper = shallow(<FilmPage {...newProps} />);
    const filmsList = wrapper.find(`FilmsList`);

    expect(filmsList.props()[`films`].length).toBeLessThanOrEqual(4);
  });

  it(`should render tab content according to the currently selected tab`, () => {
    const newProps = {
      ...props,
      location: {...props.location, hash: `#${NAV_DETAILS_ID}`}
    };

    const wrapper = shallow(<FilmPage {...newProps} />);
    expect(wrapper.find(`FilmDetails`)).toHaveLength(1);
  });

  it(`should render the first tab ("Overview") when the present location hash is not among the range of acceptable values`, () => {
    const wrapper = shallow(<FilmPage {...props} />);
    expect(wrapper.find(`FilmOverview`)).toHaveLength(1);
  });

  it(`should render the first tab ("Overview") when there's no location hash in address bar`, () => {
    const newProps = {
      ...props,
      location: {...props.location, hash: ``}
    };

    const wrapper = shallow(<FilmPage {...newProps} />);
    expect(wrapper.find(`FilmOverview`)).toHaveLength(1);
  });
});
