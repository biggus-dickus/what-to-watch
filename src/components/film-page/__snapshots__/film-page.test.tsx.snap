// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmPage test suite <FilmPage /> should render correctly 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="Bohemian Rhapsody"
          src="pic.png"
        />
      </div>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="/"
            onClick={[Function]}
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <a
            className="user-block__link"
            href="/login"
            onClick={[Function]}
          >
            Sign in
          </a>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h1
            className="movie-card__title"
          >
            Bohemian Rhapsody
          </h1>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Documentary
            </span>
            <span
              className="movie-card__year"
            >
              2019
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="img/sprite/sprite.svg#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="img/sprite/sprite.svg#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="add-review.html"
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="Bohemian Rhapsody poster"
            src="poster.jpg"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <nav
            className="movie-nav movie-card__nav"
          >
            <ul
              className="movie-nav__list"
            >
              <li
                className="movie-nav__item movie-nav__item--active"
              >
                <a
                  className="movie-nav__link"
                  href="/#overview"
                  onClick={[Function]}
                >
                  Overview
                </a>
              </li>
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  href="/#details"
                  onClick={[Function]}
                >
                  Details
                </a>
              </li>
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  href="/#reviews"
                  onClick={[Function]}
                >
                  Reviews
                </a>
              </li>
            </ul>
          </nav>
          <div
            className="movie-rating"
          >
            <div
              className="movie-rating__score"
              data-test="at-film-rating"
            >
              7.8
            </div>
            <p
              className="movie-rating__meta"
            >
              <span
                className="movie-rating__level"
                data-test="at-film-rating-word"
              />
              <span
                className="movie-rating__count"
              >
                100500
                 ratings
              </span>
            </p>
          </div>
          <div
            className="movie-card__text"
          >
            Come Tatyana with a duck, we will eat and we will fuck
            <p
              className="movie-card__director"
            >
              <strong>
                Director: 
                Alan Smithie
              </strong>
            </p>
            <p
              className="movie-card__starring"
            >
              <strong>
                Starring: 
                Ben Dover, Major Woodie, Private Parts, I. C. Wiener
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
        data-test="at-films-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          tabIndex={0}
        >
          <a
            className="small-movie-card__wrapping-link"
            href="/https://thepiratebay.org"
            onClick={[Function]}
          >
            <i
              aria-hidden={true}
              className="small-movie-card__play-btn"
            />
            <div
              className="small-movie-card__image"
            >
              <video
                height={175}
                muted={true}
                poster="img/what-we-do-in-the-shadows.jpg"
                src="https://upload.wikimedia.org/wikipedia/commons/transcoded/2/23/Dr_Strangelove_-_Official_Trailer.webm/Dr_Strangelove_-_Official_Trailer.webm.480p.vp9.webm"
                width={280}
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                What We Do in the Shadows
              </span>
            </h3>
          </a>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          tabIndex={0}
        >
          <a
            className="small-movie-card__wrapping-link"
            href="/https://thepiratebay.org"
            onClick={[Function]}
          >
            <i
              aria-hidden={true}
              className="small-movie-card__play-btn"
            />
            <div
              className="small-movie-card__image"
            >
              <video
                height={175}
                muted={true}
                poster="img/seven-years-in-tibet.jpg"
                src="https://storage.googleapis.com/coverr-main/mp4/Ants%20Walking.mp4"
                width={280}
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Seven Years in Tibet
              </span>
            </h3>
          </a>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 
          2019
           What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
